### PR TITLE
parse and handle `where T = U` predicate

### PR DESCRIPTION
**Pull request by JulianKnodt**
Updates #20041.

I took the most low-budget approach to adding equality checks (2 subtype checks in each direction), if it makes more sense to add in one check that does just equality, please let me know.

r? @varkor

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -684,6 +684,9 @@ declare_features! (
     /// Allows `cfg(target_abi = "...")`.
     (active, cfg_target_abi, "1.55.0", Some(80970), None),
 
+    /// Allows `cfg(target_abi = "...")`.
```
Comment (PatchMixolydic) [inline compiler/rustc_feature/src/active.rs:687]: Doc comment needs to be updated here
```suggestion
    /// Allows for type equality constraints in `where` clauses.
```

**Comment by fee1-dead** [comment]
I think making it an actual PredicateKind would be better. You can use `rustc_infer::infer::equate` to do this i think

**Comment by varkor** [comment]
What's the status on the [previous blockers](https://github.com/rust-lang/rust/issues/20041#issuecomment-360768487)? Have these been addressed now?

**Comment by JulianKnodt** [comment]
It's not so clear to me whether the previous blockers are completed, but given the time that lapsed, I wouldn't be surprised if they have. I still need to clean up some of the branches (since I just threw them all in last night), and will explore if this is feasible.

File: compiler/rustc_feature/src/active.rs [hunk 2]
```diff
@@ -680,6 +680,9 @@ declare_features! (
     /// Allows `#[derive(Default)]` and `#[default]` on enums.
     (active, derive_default_enum, "1.56.0", Some(86985), None),
 
+    /// Allows `where T = U` in where predicates.
+    (active, where_pred_equality, "1.56.0", Some(87533), None),
```
Comment (varkor) [inline compiler/rustc_feature/src/active.rs:684]:
```suggestion
    (active, type_equality_constraints, "1.56.0", Some(87533), None),
```

File: compiler/rustc_parse/src/parser/generics.rs [hunk 3]
```diff
@@ -284,7 +284,7 @@ impl<'a> Parser<'a> {
             }))
         // FIXME: Decide what should be used here, `=` or `==`.
         // FIXME: We are just dropping the binders in lifetime_defs on the floor here.
-        } else if self.eat(&token::Eq) || self.eat(&token::EqEq) {
```
Comment (varkor) [inline compiler/rustc_parse/src/parser/generics.rs:287]: This would be better replaced with a diagnostic suggesting the user uses `=` instead.

**Review by varkor — COMMENTED** [review]
This could do with more tests (both successful and error cases). Also, could you add tests for the cases @nikomatsakis mentions in https://github.com/rust-lang/rust/pull/22074#issuecomment-73678356?

**Comment by varkor** [comment]
Since @nikomatsakis reviewed the original PR (https://github.com/rust-lang/rust/pull/22074) that was closed, I'm going to reassign to them.

r? @nikomatsakis

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 4]
```diff
@@ -302,6 +302,13 @@ fn predicate_references_self(
             // possible alternatives.
             if data.projection_ty.substs[1..].iter().any(has_self_ty) { Some(sp) } else { None }
         }
+        ty::PredicateKind::TypeEquate(lhs, rhs) => {
+            if lhs == self_ty || rhs == self_ty {
+                Some(sp)
+            } else {
+                None
```
Comment (JulianKnodt) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:309]: TODO: this is wrong, need to walk the types and check for self

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 5]
```diff
@@ -491,6 +492,16 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
                     }
                 }
 
+                ty::PredicateKind::TypeEquate(lhs, rhs) => {
+                    match self.selcx.infcx().can_eq(obligation.param_env, lhs, rhs) {
+                        Ok(()) => ProcessResult::Changed(vec![]),
+                        Err(e) => ProcessResult::Error(FulfillmentErrorCode::CodeSubtypeError(
+                            ExpectedFound::new(false, lhs, rhs),
```
Comment (JulianKnodt) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:499]: TODO create a new fulfillment error code

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 6]
```diff
@@ -698,6 +698,17 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
 
+                    ty::PredicateKind::TypeEquate(..) => {
+                        // Errors for `TypeEquate` predicates show up as
+                        // `SelectionError::FIXME(where_pred_equality)`,
+                        // not `Unimplemented`.
+                        span_bug!(
+                            span,
+                            "type-equate requirement gave wrong error: `{:?}`",
+                            obligation
+                        )
```
Comment (JulianKnodt) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:709]: Not entirely sure what this was referencing, I suspect it's fine to leave this?

File: compiler/rustc_typeck/src/collect.rs [hunk 7]
```diff
@@ -2294,8 +2294,14 @@ fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: DefId) -> ty::GenericP
                 }))
             }
 
-            hir::WherePredicate::EqPredicate(..) => {
-                // FIXME(#20041)
+            hir::WherePredicate::EqPredicate(where_eq_pred) => {
+                let hir::WhereEqPredicate { hir_id: _, span, lhs_ty, rhs_ty } = where_eq_pred;
+                let lhs = icx.to_ty(&lhs_ty);
+                let rhs = icx.to_ty(&rhs_ty);
+                // this is a cheap approximation of equality, by checking if both are subtypes
```
Comment (JulianKnodt) [inline compiler/rustc_typeck/src/collect.rs:2301]: TODO delete old comment of me being lazy.
Here we can also do something like check that both types aren't params, which doesn't do much for now.

File: src/librustdoc/clean/mod.rs [hunk 8]
```diff
@@ -327,6 +327,7 @@ impl<'a> Clean<Option<WherePredicate>> for ty::Predicate<'a> {
             ty::PredicateKind::RegionOutlives(pred) => pred.clean(cx),
             ty::PredicateKind::TypeOutlives(pred) => pred.clean(cx),
             ty::PredicateKind::Projection(pred) => Some(pred.clean(cx)),
+            ty::PredicateKind::TypeEquate(..) => todo!(),
```
Comment (JulianKnodt) [inline src/librustdoc/clean/mod.rs:330]: TODO not entirely sure whether to make this `None`, or have something else

**Comment by nikomatsakis** [comment]
I don't think we really want this sort of where-clause. As I've come to learn more about the limits of various logics, introducing equality checks implies the ability to say things that we probably don't want to say (although I would like to support them someday, but only after we've made a transition to chalk and decided to do it). For example, being able to write `fn foo<A, B>() where A = B` would be really nifty, but we don't have the infrastructure to support that kind of equality properly, and it moves the logic from one class of logics to another (in particular we would no longer map so directly to 1st order hereditary harrop clauses).

**Comment by JulianKnodt** [comment]
That's certainly fair, equality tends to imply `!=` which seems to be quite controversial.
I was hoping to add this to improve specialization, so that you can provide `impl<T: Iterator> Foo for T where T = Map/Other<...>`, but given the additional burdens it's understandable.
In the future if/when this future is desired, this PR can be revived but I'll close it for now.

Please let me know if there is additional work elsewhere I can do to help move specialization along!

**Pull request closed without merging**